File: 945-snakes-and-ladders/snakes-and-ladders.go
func snakesAndLadders(board [][]int) int {
	n := len(board);
	target := n * n;

	// Bước 1: Chuyển ma trận 2D thành mảng 1 chiều theo kiểu Boustrophedon
	flattened := make([]int, target+1); // từ 1 đến n^2
	index := 1;
	leftToRight := true;

	for row := n - 1; row >= 0; row-- {
		if leftToRight {
			for col := 0; col < n; col++ {
				flattened[index] = board[row][col];
				index++;
			}
		} else {
			for col := n - 1; col >= 0; col-- {
				flattened[index] = board[row][col];
				index++;
			}
		}
		leftToRight = !leftToRight;
	}

	// Bước 2: BFS với hàng đợi
	queue := make([]int, target); // dùng như hàng đợi vòng
	head, tail := 0, 0;
	queue[tail] = 1;
	tail++;

	steps := make([]int, target+1); // steps[i] = số bước để đến ô i
	steps[1] = 1;

	for head != tail {
		pos := queue[head];
		head = (head + 1) % target;

		// Nếu trong 1 roll tới được đích thì trả kết quả luôn
		if pos+6 >= target {
			return steps[pos];
		}

		maxNeutral := 0;

		for roll := 6; roll >= 1; roll-- {
			next := pos + roll
			if next > target {
				continue;
			}

			if flattened[next] != -1 {
				// Có snake hoặc ladder
				next = flattened[next];
				if next == target {
					return steps[pos];
				}
			} else {
				// Không có snake/ladder
				if roll < maxNeutral {
					continue;
				}
				maxNeutral = roll;
			}

			if steps[next] == 0 {
				steps[next] = steps[pos] + 1;
				queue[tail] = next;
				tail = (tail + 1) % target;

				// Nếu head == tail thì hàng đợi bị tràn
				if head == tail {
					return 0;
				}
			}
		}
	}

	return -1; // Không thể đến đích
}

File: 945-snakes-and-ladders/snakes-and-ladders.java
class Solution {
    public int snakesAndLadders(int[][] board) {
        int size = board.length; // Kích thước bảng (n)
        int target = size * size; // Vị trí đích là n^2

        // Bước 1: Chuyển bảng 2D thành mảng 1 chiều theo thứ tự Boustrophedon
        short[] flattened = new short[target + 1]; // flattened[1..n^2]
        int index = 1;

        for (int row = size - 1; row >= 0; row--) {
            // Nếu dòng này đi từ trái sang phải
            for (int col = 0; col < size; col++)
                flattened[index++] = (short) board[row][col];

            // Kiểm tra nếu còn dòng tiếp theo không
            if (--row < 0)
                break;

            // Dòng kế tiếp đi từ phải sang trái
            for (int col = size - 1; col >= 0; col--)
                flattened[index++] = (short) board[row][col];
        }

        // Bước 2: Chuẩn bị BFS với hàng đợi vòng
        short[] queue = new short[target]; // Hàng đợi vòng để tiết kiệm bộ nhớ
        int head = 0, tail = 0;
        queue[tail++] = 1; // Bắt đầu từ ô số 1

        int[] steps = new int[target + 1]; // steps[i] là số bước cần để đến ô i
        steps[1] = 1; // Bắt đầu từ bước 1 (tức 0 lần tung xúc xắc + 1)

        // Bước 3: Thực hiện BFS
        while (head != tail) {
            int position = queue[head++];
            head %= target; // Đảm bảo head không vượt khỏi mảng

            // Nếu trong 1 nước đi tới được đích thì trả về kết quả ngay
            if (position + 6 >= target)
                return steps[position];

            int maxNeutral = 0; // Lưu roll cao nhất không bị snake/ladder để tránh duplicate

            // Thử các giá trị xúc xắc từ 6 đến 1
            for (int roll = 6; roll >= 1; roll--) {
                int next = position + roll;

                if (flattened[next] >= 0) {
                    // Có snake/ladder => buộc phải nhảy đến đích của nó
                    next = flattened[next];
                    if (next == target)
                        return steps[position]; // Về đích
                } else {
                    // Nếu không có snake/ladder
                    if (roll < maxNeutral)
                        continue; // Bỏ nếu không phải roll cao nhất
                    maxNeutral = roll;
                }

                // Nếu ô này chưa từng được đến thì thêm vào hàng đợi
                if (steps[next] == 0) {
                    steps[next] = steps[position] + 1;
                    queue[tail++] = (short) next;
                    tail %= target;

                    // Nếu hàng đợi bị tràn (circular queue đầy) => lỗi logic
                    if (head == tail)
                        return 0;
                }
            }
        }

        // Không thể đến ô cuối cùng
        return -1;
    }
}
